crates/red_knot/src/lint.rs

```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
     Ok(())
 }
 
+fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
+    // TODO possibly we should have a special marker on the real typing module (from typeshed) so
```

Comment:
Yes, mypy special-cases a small number of stdlib modules (including `typing`) such that if one of them is shadowed, that's a hard error and mypy will refuse to do any type-checking at all. I think it would make sense for us to do similarly. See e.g. https://github.com/python/mypy/pull/13155

We may also want a more general way to be able to distinguish between "`datetime.datetime`, where `datetime` is the stdlib `datetime` module", and "`datetime.datetime`, where `datetime` is a first-party user module" though. For a type checker that distinction doesn't matter so much — as long as it can follow the module resolution and figure out the type, it's all good. But we probably _don't_ want to e.g. emit our `flake8-datetimez` rules on code using functions from random user modules that shadow the stdlib `datetime` module. We probably _only_ want to emit those lines on code using functions imported from the _actual stdlib_ `datetime` module.